packages/react/src/ReferenceRangeEditor/ReferenceRangeEditor.tsx

```diff
@@ -0,0 +1,439 @@
+  Text,
+  TextInput,
+} from '@mantine/core';
+import { CodeableConcept, ObservationDefinition, ObservationDefinitionQualifiedInterval } from '@medplum/fhirtypes';
+import { IconCircleMinus, IconCirclePlus } from '@tabler/icons';
+import { killEvent } from '../utils/dom';
+
+import { formatRange, getCodeBySystem } from '@medplum/core';
+import React, { useEffect, useState } from 'react';
+import { Form } from '../Form/Form';
+import { RangeInput } from '../RangeInput/RangeInput';
+
+const useStyles = createStyles((theme) => ({
+  section: {
+    position: 'relative',
+    margin: '4px 4px 8px 0',
+    padding: '6px 12px 16px 6px',
+    border: `1.5px solid ${theme.colors.gray[3]}`,
+    borderRadius: theme.radius.sm,
+    transition: 'all 0.1s',
+  },
+}));
+
+// Properties of qualified intervals used for grouping
+const intervalFilters = ['gender', 'age', 'gestationalAge', 'context', 'appliesTo'] as const;
+export interface ReferenceRangeEditorProps {
+  definition: ObservationDefinition;
+  onSubmit: (result: ObservationDefinition) => void;
+}
+
+// Helper type that groups of qualified intervals by equal filter criteria
+type IntervalGroup = {
+  id: string;
+  filters: Record<typeof intervalFilters[number], any>;
+  intervals: ObservationDefinitionQualifiedInterval[];
+};
+
+const defaultProps: ReferenceRangeEditorProps = {
+  definition: { resourceType: 'ObservationDefinition' },
+  onSubmit: () => {
+    return;
+  },
+};
+
+export function ReferenceRangeEditor(props: ReferenceRangeEditorProps): JSX.Element {
+  props = Object.assign(defaultProps, props);
+  const defaultDefinition = props.definition;
+
+  const [intervalGroups, setIntervalGroups] = useState<IntervalGroup[]>([]);
+  const [groupId, setGroupId] = useState(1);
+  const [intervalId, setIntervalId] = useState(1);
+
+  useEffect(() => {
+    const definition = ensureQualifiedIntervalKeys(defaultDefinition, setIntervalId);
+    setIntervalGroups(groupQualifiedIntervals(definition.qualifiedInterval || [], setGroupId));
+  }, [defaultDefinition]);
+
+  return (
+    <Form testid="reference-range-editor" onSubmit={submitDefinition}>
+      <Stack>
+        {intervalGroups.map((intervalGroup) => (
+          <ReferenceRangeGroupEditor
+            unit={getUnitString(defaultDefinition.quantitativeDetails?.unit)}
+            onChange={changeInterval}
+            onAdd={addInterval}
+            onRemove={removeInterval}
+            onRemoveGroup={removeGroup}
+            key={`group-${intervalGroup.id}`}
+            intervalGroup={intervalGroup}
+          />
+        ))}
+      </Stack>
+      <ActionIcon
+        title="Add Group"
+        size="sm"
+        onClick={(e: React.MouseEvent) => {
+          killEvent(e);
+          addGroup({ id: `group-id-${groupId}`, filters: {} as IntervalGroup['filters'], intervals: [] });
+          setGroupId((id) => id + 1);
+        }}
+      >
+        <IconCirclePlus />
+      </ActionIcon>
+
+      <Button type="submit">Save</Button>
```

Comment:
For "end of form" buttons, I've been trying to use:

```tsx
<Group position="right">
 <Button type="submit">Save</Button>
</Group>
```

or, if you need more vertical spacing:

```tsx
<Group position="right" mt="xl">
 <Button type="submit">Save</Button>
</Group>
```